programs/survival/what_biome_on_block.sc

```diff
@@ -0,0 +1,79 @@
+// shows biomes blockwise. useful for marking out rivers, beaches, soul sand valley and stuff.
+// now works across diemsions (yes even in the_end)
+__config()->{
+    'stay_loaded'->true,
+    'scope' -> 'global',
+    'commands' -> {
+        '' -> _() -> print(format('r Missing arguments. Correct syntax /showbiomes <toggle>')),
+        '<toggle>' -> l('__toggle_markings'),
+        'print' -> _() -> print(global_players)
+    },  
+    'arguments' -> {
+        'toggle' -> {'type'->'text','options' -> l('on', 'off')},
+    }
+};
+
+// if the script crashes your multiplayer world, do one of the two
+// 1. update carpet to latest version
+// 2. turn this to true 
+global_carpet_crash_fix = false;
+
+// change this global_radius to improve fps if you on a potato. 4 works decently
+global_radius = 4;
+
+// change the font size to your liking
+global_font_size = 6;
+
+// dont change anything this point onward
+global_players = m();
+global_dimension_map = m('overworld' -> 'foliage_color', 'the_nether' -> 'fog_color', 'the_end' -> 'sky_color');
+
+__make_markers(player) -> (
+    l(posx, posy, posz)  = pos(player);
+    scan(posx, posy, posz, -global_radius, 0, global_radius,
+        l(x, y, z) = pos(_);
+        draw_shape(
+            'box', 
+            2,
+            'color', 0x00000000,
+            'fill', if(global_carpet_crash_fix, 0x808080FF, biome(_, global_players:player:'dim_color')) - 0x0000009F, 
+            'from', l(x, y + 0.05, z), 
+            'to', l(x + 1, y - 0.05, z + 1),
+        );
+
+        draw_shape(
+            'label', 
+            2,
+            'color', 0xFFFFFFFF,
+            'text', biome(_),
+            'pos', l(x + 0.5, y, z + 0.5),
+            'size', global_font_size,
+        );
+    );
+);
+
+__add_player(player) -> (
+    global_players = global_players + {player -> {'dim_color' -> global_dimension_map:(player ~ 'dimension'), 'toggle' -> false}};
+);
+
+__toggle_markings(toggle) -> (
+    p = player();
+    if(!global_players:p, __add_player(p));
+    global_players:p:'toggle' = toggle == 'on';
+);
+
+__on_player_connects(player) -> (
+    __add_player(player);
+);

```

Comment:
If you don't make the app global scoped, you can remove this. Only reason I can see to make it global is that `print` shows the current setting for all online players, instead of just the caller? idk if that's intended.

Reply:
Print was the debug command left behind. I will update it.

Reply:
Also, is there a upside/downside of loading as global ? If there's none, I can for sure get rid of the global list.

Reply:
Loading app as global means there is a single instance of the app, sharing the same variables and everything else.

Player scope means there is an instance for every player, each with their own globals (and handlers I think?).

The overhead of one or the other is not noticeable, if any.

Reply:
Isn't a single instance shared better than individual instances for all players ? Also if theres no upside of not using player scope, I will stick with global scope.

Reply:
The upside is that variable handling is easier, since you don't ahve to keep a map indexing your settings for each player. Just a `global_state` will suffice, no need to query current player. Gnembon also sometimes says they are "safer", I think he referes to the fact that you don't get bugs from one player changing something for another player where the programmer didn't forsee it being possible.

Reply:
This way you don't need to concern yourself with a `global_players` map and the corresponding `__on_player_connects` and `-disconnects` events and stuff like that.